include/gsl/gsl

```diff
@@ -74,69 +74,65 @@ class not_null
     static_assert(std::is_assignable<T&, std::nullptr_t>::value, "T cannot be assigned nullptr.");
 
 public:
-    not_null(T t) : ptr_(t) { ensure_invariant(); }
-    not_null& operator=(const T& t)
-    {
-        ptr_ = t;
-        ensure_invariant();
-        return *this;
-    }
+    constexpr not_null(T t) : ptr_(t) { ensure_invariant(); }
 
     not_null(const not_null& other) = default;
     not_null& operator=(const not_null& other) = default;
 
     template <typename U, typename Dummy = std::enable_if_t<std::is_convertible<U, T>::value>>
-    not_null(const not_null<U>& other)
+    constexpr not_null(const not_null<U>& other) : ptr_(other.get())
     {
-        *this = other;
+        ensure_invariant(); // Just in case a non-null U converts to null T.
     }
 
-    template <typename U, typename Dummy = std::enable_if_t<std::is_convertible<U, T>::value>>
-    not_null& operator=(const not_null<U>& other)
+    constexpr T get() const
     {
-        ptr_ = other.get();
-        return *this;
+        // help the optimizer
+        GSL_ASSUME(ptr_ != nullptr);
```

Comment:
I appreciate the aim to aid the optimizer. But could we combine this with also testing/indicating the contract on the function? How about replacing the `GSL_ASSUME()` with an `Ensures()` and then updating `Ensures()` (for the relevant platforms) to use `__assume()`?